File: ios/ClingSync/UploadReducer.swift
import Foundation

// Pure transition table for the upload state machine, operating on plain data so
// it is exhaustively unit-testable without the upload mechanism or any IO. Ports
// Android's UploadReducer, including the guards that keep already-finished rows
// from being reverted on abort/failure.
enum UploadReducer {
    static func reduce(_ state: AppState, _ update: WorkUpdate) -> AppState {
        switch update {
        case .enqueued(let id):
            var next = state
            next.currentUploadId = id
            return next
        case .running(let statuses, let uploadedBytes):
            return running(state, statuses: statuses, uploadedBytes: uploadedBytes)
        case .succeeded(let finalStatuses):
            return succeeded(state, finalStatuses: finalStatuses)
        case .failed(let error):
            return failed(state, error: error)
        case .cancelled:
            return cancelled(state)
        }
    }

    private static func running(
        _ state: AppState,
        statuses: [String: FileStatus],
        uploadedBytes: Int64
    ) -> AppState {
        // Ignore a progress signal with nothing to attach it to: a stale "running"
        // after the upload ended (the abort race). A reattached upload after
        // process death has currentUploadId set (idle() nulls it on end), so it is
        // correctly NOT dropped here.
        if !state.isUploading, !state.isUploadInitiated, state.currentUploadId == nil {
            return state
        }

        // A transient unreadable status file yields no statuses; flag uploading but
        // keep the previous uploadInfo rather than resetting it to 0/0.
        if statuses.isEmpty {
            var next = state
            next.isUploading = true
            return next
        }

        let uploadingId = statuses.first { $0.value == .sending }?.key
        let currentFileName = uploadingId.flatMap { id in state.files.first { $0.id == id }?.name }
        let total = statuses.count
        let completed = statuses.values.filter { isCompleted($0) }.count
        let anyCommitting = statuses.values.contains { $0 == .committing }
        let info: UploadInfo
        if anyCommitting {
            info = UploadInfo(currentFile: "Committing changes...", currentIndex: completed, totalFiles: total)
        } else {
            info = UploadInfo(currentFile: currentFileName, currentIndex: completed, totalFiles: total)
        }

        var next = state
        next.isUploading = true
        next.fileStatus.merge(statuses) { _, new in new }
        next.uploadInfo = info
        next.uploadedBytes = uploadedBytes
        return next
    }

    private static func succeeded(_ state: AppState, finalStatuses: [String: FileStatus]) -> AppState {
        var next = idle(state)
        next.fileStatus.merge(finalStatuses) { _, new in new }
        next.selectedIds = []
        return next
    }

    private static func failed(_ state: AppState, error: String) -> AppState {
        var next = idle(state)
        next.fileStatus = markUnfinished(state, status: .failed(message: "Error"))
        if state.overlay == .none {
            next.overlay = .error(title: "Upload Failed", message: error)
        }
        return next
    }

    private static func cancelled(_ state: AppState) -> AppState {
        var next = idle(state)
        next.fileStatus = markUnfinished(state, status: .aborted)
        return next
    }

    private static func isCompleted(_ status: FileStatus) -> Bool {
        switch status {
        case .sentWaitingCommit, .committing:
            return true
        case .exists:
            return true
        default:
            return false
        }
    }

    // A file whose terminal status must survive an abort/failure: already uploaded
    // (and committed elsewhere in the batch) or already present in the repo.
    private static func keepsTerminalStatus(_ status: FileStatus?) -> Bool {
        switch status {
        case .done, .exists:
            return true
        default:
            return false
        }
    }

    // Mark every in-flight file (not already Done/Exists) with the given status.
    private static func markUnfinished(_ state: AppState, status: FileStatus) -> [String: FileStatus] {
        var result = state.fileStatus
        for id in state.currentUploadIds where !keepsTerminalStatus(state.fileStatus[id]) {
            result[id] = status
        }
        return result
    }

    // Reset every per-upload field back to the not-uploading baseline.
    private static func idle(_ state: AppState) -> AppState {
        var next = state
        next.isUploading = false
        next.isUploadInitiated = false
        next.currentUploadId = nil
        next.currentUploadIds = []
        next.uploadInfo = nil
        next.uploadedBytes = 0
        return next
    }
}

File: android/app/src/main/java/com/clingsync/android/presentation/UploadReducer.kt
package com.clingsync.android.presentation

import com.clingsync.android.FileStatus
import com.clingsync.android.UploadInfo
import java.io.File

// Pure transition table for the WorkManager-driven upload state machine. Every
// rule that used to live in MainActivity's LaunchedEffect(uploadWorkInfos) is
// here, operating on plain data so it is exhaustively unit-testable without
// WorkManager, an emulator, or any IO.
object UploadReducer {
    fun reduce(
        state: MainUiState,
        update: WorkUpdate,
    ): MainUiState =
        when (update) {
            is WorkUpdate.Enqueued -> state.copy(currentUploadId = update.id)
            is WorkUpdate.Running -> running(state, update)
            is WorkUpdate.Succeeded -> succeeded(state, update)
            is WorkUpdate.Failed -> failed(state, update)
            WorkUpdate.Cancelled -> cancelled(state)
        }

    private fun running(
        state: MainUiState,
        update: WorkUpdate.Running,
    ): MainUiState {
        // Ignore a progress signal when there is no upload to attach it to: a
        // stale "Running" after the upload ended (the abort race the old code
        // had). A reattached upload after process death has currentUploadId set
        // (idle() nulls it on end), so it is correctly NOT dropped here.
        if (!state.isUploading && !state.isUploadInitiated && state.currentUploadId == null) return state

        // A transient unreadable status file yields no statuses; flag uploading
        // but keep the previous uploadInfo rather than resetting it to 0/0.
        if (update.statuses.isEmpty()) return state.copy(isUploading = true)

        val uploadingPath = update.statuses.entries.firstOrNull { it.value is FileStatus.Uploading }?.key
        val currentFileName = uploadingPath?.let { File(it).name }
        val completed =
            update.statuses.count {
                it.value is FileStatus.Uploaded || it.value is FileStatus.Exists || it.value is FileStatus.Committing
            }
        val total = update.statuses.size
        val info =
            when {
                update.statuses.any { it.value is FileStatus.Committing } ->
                    UploadInfo("Committing changes...", completed, total)
                currentFileName != null -> UploadInfo(currentFileName, completed, total)
                else -> UploadInfo(null, completed, total)
            }
        return state.copy(
            isUploading = true,
            fileStatus = state.fileStatus + update.statuses,
            uploadInfo = info,
            uploadedBytes = update.uploadedBytes,
        )
    }

    private fun succeeded(
        state: MainUiState,
        update: WorkUpdate.Succeeded,
    ): MainUiState =
        idle(state).copy(
            fileStatus = state.fileStatus + update.finalStatuses,
            selectedPaths = emptySet(),
        )

    private fun failed(
        state: MainUiState,
        update: WorkUpdate.Failed,
    ): MainUiState =
        idle(state).copy(
            fileStatus = markUnfinished(state, FileStatus.Failed("Error")),
            // The share surfaces a failure through its own outcome dialog (which returns
            // to the main app), so it must not also raise the error overlay.
            overlay =
                if (!state.shareMode && state.overlay is Overlay.None) {
                    Overlay.Error("Upload Failed", update.error)
                } else {
                    state.overlay
                },
        )

    private fun cancelled(state: MainUiState): MainUiState = idle(state).copy(fileStatus = markUnfinished(state, FileStatus.Aborted))

    // Mark every in-flight file (not already Done/Exists) with the given status.
    private fun markUnfinished(
        state: MainUiState,
        status: FileStatus,
    ): Map<String, FileStatus> {
        val updates =
            state.currentUploadPaths
                .filter { state.fileStatus[it] !is FileStatus.Done && state.fileStatus[it] !is FileStatus.Exists }
                .associateWith { status }
        return state.fileStatus + updates
    }

    // Reset every per-upload field back to the not-uploading baseline.
    private fun idle(state: MainUiState): MainUiState =
        state.copy(
            isUploading = false,
            isUploadInitiated = false,
            currentUploadId = null,
            currentUploadPaths = emptySet(),
            uploadInfo = null,
            uploadedBytes = 0L,
        )
}
